docs/backend/backend-overview.md

```diff
@@ -0,0 +1,209 @@
+focused on evaluating taxonomy contributions and running batches of them through
+the backend pipeline.
+
+![High-Level Service Diagram](./images/backend-service-high-level.png)
+
+### Observable
+
+The Backend Service must be observable, both by the Backend Service Owners and
+by InstructLab Model Developers.
+
+### Reusable
+
+The Backend Service was envisioned from the beginning as a pipeline of distinct
+steps (data generation, training, evaluation, and publishing). This separation
+of concerns should be maintained in a future service-oriented architecture. While
+the diagram under [Service-Oriented](#service-oriented) shows a high-level diagram
+with a single "InstructLab Service API," in practice, this should be broken down
+into multiple services with a more focused set of concerns.
+
+Within InstructLab, we could have a top-level application that implements the
+entire backend pipeline workflow and does so by making use of each of these
+services. However, in the future, we could also envision these services being
+used in other contexts.
+
+![Service Breakdown](./images/backend-service-breakdown.png)
+
+### Small Iterations
+
+While we may cast a vision for the future, we desire to get there via a series
+of small iterations. We have a working backend pipeline already and will
+continue to run it every week. In parallel, we can work to evolve it through a
+series of small iterations that evolve our processes over time to use new
+service-based interfaces, gradually evolving the execution process that occurs
+each week.
+
+We do not want a new effort happening in parallel to what is already working
+today.
+
+### Minimize Mandatory External Dependencies
+
+Major external dependencies should be made optional where feasible to do so.
+The more of these we require, the more complicated it will be to reuse this
+system in other environments.
+
+In other cases, it may be appropriate to provide an abstraction for the
+dependency so that alternatives can be put in place. Alternate Object Storage
+systems are a good example.
+
+### Make Development Accessible
+
+Most people will not have direct access to a large cluster to test changes
+to this system. We should create development environments and minimal
+configurations as much as possible to lower the barrier to doing useful
+development and testing with these components.
+
+### Model Independence
+
+We desire to keep the backend flexible such that serving and training alternate
+models in the future is not more work than absolutely necessary.
+
+## Key Technology Decisions
+
+### Kubernetes
+
+We consider Kubernetes an acceptable dependency for the Backend Service. While
+it may be technically possible to run most of the code elsewhere, the project
+will focus its energy on running this collection of components in a
+Kubernetes-based environment.
+
+This also applies to when these components are run on a single server. In that
```

Comment:
I would prefer putting that off into a future, dedicated proposal. The simplest way to get started is to focus on clustered operation, as that is where the "real" backend runs anyway. That is what we're aiming to replace as quick as possible.

Reply:
Sure, I see this as version 1.1 or 2.0 of the system. The key idea is we need to get to the point that `ilab command` on linux talks to the RESTAPI for launching commands. On RHEL AI, we can back the RESTAPI by launching podman containers or talking to podman services running on the system, on OpenShift AI systems, the RESTAPI can launch kubernetes workloads.

When we get a plugin for Podman Desktop, it could launch containers in the local podman machine or talk directly to the RESTAPI on RHEL AI or OpenShift AI.

Reply:
Yeah, I understand the proposal. I just prefer to keep it out of scope for now.

Reply:
@russellb is the goal of keeping non-k8s REST APIs out of scope to make inclusion of single-node k8s the intentional direction for RHEL AI? That seems like a large discussion that may be bigger than this review.

Reply:
Its ok to keep it out of scope as long as the backend can be deployed independent of containers and k8s cluster. It can be flexible.